oshdb-util/src/main/java/org/heigit/ohsome/oshdb/util/osh/OSHEntityTimeUtils.java

```diff
@@ -0,0 +1,381 @@
+      OSHEntity osh, Predicate<OSMEntity> osmEntityFilter) {
+    List<OSHDBTimestamp> result = new ArrayList<>();
+    // `nextNonMatchTime` stores the timestamp when the entity is deleted or doesn't match the
+    // filter anymore. This is used to fold consecutive "non-matching" versions. example:
+    //     a way has tag K=A in version 1, tag K=B in version 2 and is deleted in version 3, and
+    //     the filter only matches versions with the tag K=A. Then in the first iteration of the
+    //     loop, the nextNonMatchTime will be set to t3, in the second iteration it will be set
+    //     to t2 and in the third iteration it will be added to the result.
+    OSHDBTimestamp nextNonMatchTime = null;
+    for (OSMEntity osm : osh.getVersions()) {
+      if (osm.isVisible() && (osmEntityFilter == null || osmEntityFilter.test(osm))) {
+        if (nextNonMatchTime != null) {
+          // the next version of this entity is deleted or doesn't match the filter anymore
+          // -> also return the time of the "deletion"
+          result.add(nextNonMatchTime);
+          nextNonMatchTime = null;
+        }
+        result.add(osm.getTimestamp());
+      } else {
+        // save the time of the next "deletion" for later
+        nextNonMatchTime = osm.getTimestamp();
+      }
+    }
+    return result;
+  }
+
+  /**
+   * Puts all changeset timestamps into the result of the given OSH entity.
+   */
+  private static void putChangesetTimestamps(OSHEntity osh, Map<OSHDBTimestamp, Long> result) {
+    for (OSMEntity osm : osh.getVersions()) {
+      result.putIfAbsent(osm.getTimestamp(), osm.getChangesetId());
+    }
+  }
+
+  /**
+   * Puts all changeset timestamps into the result of all OSH nodes in the given list.
+   */
+  private static void putChangesetTimestamps(
+      List<OSHNode> oshList, Map<OSHDBTimestamp, Long> result) {
+    for (OSHEntity osh : oshList) {
+      if (osh == null) {
+        continue;
+      }
+      for (OSMEntity osm : osh.getVersions()) {
+        result.putIfAbsent(osm.getTimestamp(), osm.getChangesetId());
+      }
+    }
+  }
+
+  /**
+   * Puts all changeset timestamps into the result of all OSH entities in the given list.
+   */
+  private static void putChangesetTimestampsRecurse(
+      List<? extends OSHEntity> oshList, Map<OSHDBTimestamp, Long> result) {
+    for (OSHEntity osh : oshList) {
+      if (osh == null) {
+        continue;
+      }
+      getChangesetTimestamps(osh).forEach(result::putIfAbsent);
+    }
+  }
+
+  /**
+   * Returns for each member of an entity a list of time intervals when the belong to the entity
+   * matching the given predicate.
+   */
+  private static Map<OSHEntity, LinkedList<OSHDBTimestamp>> collectMembershipTimeIntervals(
+      OSHEntity osh, Predicate<OSMEntity> osmEntityFilter) {
+    Map<OSHEntity, LinkedList<OSHDBTimestamp>> memberTimes
+        = new TreeMap<>(Comparator.comparingLong(OSHEntity::getId));
```

Comment:
A TreeMap is not requiered at this point, so a HashMap would fit better here.

Reply:
Am I right that the benefit of `HashMap` (over other `Map` implementations) would be faster access when calling `.get` below?

Reply:
That's true. for a tree map you have to do potential more checks than with a HashMap.
We should also keep in mind, that it could be worth, to check if the members actually change between one major-version to the next, I'm not sure how much of an impact this has. Could be something for the next PR.

Reply:
PS: Could it be an issue for using a HashMap that OSHEntity does not implement hashCode and equals?

>  We should also keep in mind, that it could be worth, to check if the members actually change between one major-version to the next, I'm not sure how much of an impact this has. Could be something for the next PR.

I'm not 100% sure if I understand what you propose, but this method does indeed not look at what happens to the members, it just collects the time intervals where they are a member of the entity at hand. The actual _change_s (modificationTimestamps) are calculated in `fillMembersModificationTimestamps` below.

Reply:
> Could it be an issue for using a HashMap that OSHEntity does not implement hashCode and equals?
Yes, very good point, that's could be an issue, normally! 
Every object inherit the hashcode, equals from the Object class, which do a check an the "objects address". 
But in this case the returned  osh-entity from member.getEntity() is always the same object.

Reply:
Actually, good that we think about this
```java
Map<OSHEntity, LinkedList<OSHDBTimestamp>> memberTimes
        = new TreeMap<>(Comparator.comparingLong(OSHEntity::getId));
```

is a potential bug in case the osh argument of ` collectMembershipTimeIntervals` is of type relation!
It is possible that ( and I know that here are cases), that a relation has a way member with the same id with a node member.

Reply:
you are right! :scream: good catch! I will switch to the HashMap as suggested.